### PR TITLE
[ramda] Add type predicate overload to ifElse

### PR DESCRIPTION
**Pull request by JeongJuhyeon**
Please fill in this template.

- [x]  Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [ ] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change. **Tests failed before this PR and still fail in the same way. Fixing them is outside of the scope of this PR. All it does is add an overload.**
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <<https://github.com/ramda/ramda/issues/3215>>

### CODE REVIEW

**Comment by DangerBotOSS** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## ramda ([<kbd>unpkg</kbd>](https://unpkg.com/browse/ramda@latest/))
was missing the following properties:
1. sequence
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 0f081badfca03d52ac5c903dbfa71d097df95f00
</p>

**Comment by adispring** [comment]
Can you add some tests about this change ?

**Comment by JeongJuhyeon** [comment]
As I mentioned in the PR, the existing test  already didn't compile on the master brancg and I'm not sure what the best way would be to fix it. So even if I add a test it can't be ran. Do you still want me to add one anyway?

Android용 Outlook<https://aka.ms/ghei36> 받기

________________________________

**Comment by JeongJuhyeon** [comment]
I'm a bit confused; my minimal example behaves differently when using ramda as a library vs. when writing it in the test. This is my first time contributing to DefinitelyTyped so maybe this is a common differencethat I'm not aware of.

`yarn add typescript ramda @types/ramda`

```typescript
type stringOrNumber = string | number;
const isNumber = (o: stringOrNumber): o is number => typeof o === 'number';
function takeNumber(x: number) {}
function takeString(s: string) {}

/* Error: Argument of type '(s: string) => void' is not assignable to parameter of type '(x: number) => void'.
  Types of parameters 's' and 'x' are incompatible.
  Type 'number' is not assignable to type 'string' */
// Inferred type: function ifElse<[x: number], void, void>(fn: (x: number) => boolean, onTrue: (x: number) => void, onFalse: (x: number) => void): (x: number) => void
R.ifElse(isNumber, takeNumber, takeString);
```

But when pasting the exact same code into the ifElse test in the DefinitelyTyped repo:

```typescript
// Inferred type: function ifElse<[o: stringOrNumber], void, void>(fn: (o: stringOrNumber) => boolean, onTrue: (o: stringOrNumber) => void, onFalse: (o: stringOrNumber) => void): (o: stringOrNumber) => void
```

File: types/ramda/index.d.ts [hunk 1]
```diff
@@ -888,6 +888,7 @@ export function identity<T>(a: T): T;
  */
 // tslint:disable:max-line-length
 export function ifElse<TArgs extends any[], TOnTrueResult, TOnFalseResult>(fn: (...args: TArgs) => boolean, onTrue: (...args: TArgs) => TOnTrueResult, onFalse: (...args: TArgs) => TOnFalseResult): (...args: TArgs) => TOnTrueResult | TOnFalseResult;
+export function ifElse<TArgs extends any, N extends TArgs, TOnTrueResult, TOnFalseResult>(fn: (args: TArgs) => args is N, onTrue: (args: N) => TOnTrueResult, onFalse: (args: Exclude<TArgs, N>) => TOnFalseResult): (args: TArgs) => TOnTrueResult | TOnFalseResult;
```
Comment (adispring) [inline types/ramda/index.d.ts:891]: `ifElse` support multiple arguments, but this function type only has one argument.

Why `onFalse` must exclude N ? The type should be common enough to cover most case.

**Comment by JeongJuhyeon** [comment]
@adispring 

>Why onFalse must exclude N ? The type should be common enough to cover most case.

Because that's the point of a [type predicate](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates). It takes an argument which is a union and calling it upon a variable of that union cause its type to be narrowed. We have a variable that is A | B | C | D, if the predicate is true the type is narrowed to A, otherwise it is B | C | D. If the predicate is false then there will (and should) be an error when trying to use the variable as an argument for a function that takes A. 
The TS compiler itself excludes the narrowed type from the union in the else clause. Here's a [TS Playground example](https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABGEBbARgUwE4HEQCG2AJgBQBWAXMmltogD6IDOU2MYA5o4gN4DaAa0wBPaq3ZcAutQJgRAXwCU1cohjMaGHHwBQiA4myYoIbEigiADpjjBEagLzPEAchTbsrgNy6Fu0EhYBEQoAmEAOVocUgAPag86JV5-APBoeCQQZkx8IjIqLToeCQ5uJgFhMRY2MplEOUUlPUN1e1JEnDySCiVm3n1WwwgEVkRBREcHXyHDMMjo7FJBJRnDBURMABscltnEEbAxianyb0QAeguD0ahx8VquHkrRB8lOesb-fdDwzCjPMtVoMDP4FEA).

Due to the first argument being `fn: (args: TArgs) => args is N`, and the overload being below the first one, it will only be matched on when the first argument ifElse is being called with is explicitly a type predicate, in which case the above is the desired behavior. If the first argument is not a type predicate, the overload will not be matched on, so I don't see any risks of undesired behaviour.

My [original issue](https://github.com/ramda/ramda/issues/3215) also has a code example.

>ifElse support multiple arguments, but this function type only has one argument.

In Typescript, type predicates only support narrowing the type of a single variable. Since the role of the type predicate is to narrow the type of a single variable, having it take multiple arguments is likely an antipattern. The type of a variable by definition does not depend on other variables, just like the number of bytes it takes up in memory.

It's reasonable to suggest that the branches (onTrue and onFalse) should be able to take other arguments too. However as mentioned, for there to be a point in branching on a type predicate, we need to use its narrowed result which leaves us with no option (unless you have any workarounds!) other than to have both branches take it as their first argument. I've updated the PR to include this. 1. Type predicates are an important feature of Typescript 2. They're perfect to branch on 3. Currently they cannot be used with R.ifElse  (and similar Ramda functions) at all. Again, since it's only an overload, and will only ever be matched on if the first argument is not a type predicate **plus** the existing type definition doesn't match, I think it's a reasonable addition with no risk of breaking anyone's code.

File: types/ramda/index.d.ts [hunk 2]
```diff
@@ -888,6 +888,7 @@ export function identity<T>(a: T): T;
  */
 // tslint:disable:max-line-length
 export function ifElse<TArgs extends any[], TOnTrueResult, TOnFalseResult>(fn: (...args: TArgs) => boolean, onTrue: (...args: TArgs) => TOnTrueResult, onFalse: (...args: TArgs) => TOnFalseResult): (...args: TArgs) => TOnTrueResult | TOnFalseResult;
+export function ifElse<TArgs extends any[], U extends any, N extends U, TOnTrueResult, TOnFalseResult>(fn: (u: U) => u is N, onTrue: (n: N, ...args: TArgs) => TOnTrueResult, onFalse: (n: Exclude<U, N>, ...args: TArgs) => TOnFalseResult): (u: U, ...args: TArgs) => TOnTrueResult | TOnFalseResult;
```
Comment (amcasey) [inline types/ramda/index.d.ts:891]: Is `TArgs` still needed?  Is it just there to support the case where the predicate takes fewer arguments than `onTrue` and `onFalse`?

What does the `extends any` in `U extends any` do?

**Review by amcasey — CHANGES_REQUESTED** [review]
I'm a little concerned that the new overload addresses a very narrow use case in a widely used library and there doesn't seem to be a lot of support from other users or the library maintainers.  Is this issue more prevalent than I realize?

Remaining concerns @adispring?

**Pull request closed without merging**